Add `port`, `gdb_args`, and `gdbserver_args` to gdb.debug()

gfelber opened the pull request:
added two arguments to gdb.debug()
+ port: specifies the port that should be used by the gdbserver, this is useful for alpine, because alpine doesn't support the default openssh port forwarding feature, due to a minimalistic version of openssh
+ gdb_args: allows forwarding arguments to the gdb binary spawned by gdb.attach() (just a passthrough)

also downgraded "GDB Python API is supported only for local processes" to a warning, because it seems to work.

These changes are made for compatibility with my pwntools "plugin" [vagd](https://github.com/gfelber/vagd) which currently uses a workaround on [Pwngd.debug()](https://github.com/gfelber/vagd/blob/67dcdfc08e04144ecf4abf7a661793e627e041b3/src/vagd/virts/pwngd.py#L204) (basically the same behaviour as the patch)

Commit:
added two arguments to gdb.debug:
+ port: specifies the port that should be used by the gdbserver (this is useful for alpine)
+ gdb_args: allows forwarding arguments to the gdb binary spawned by gdb.attach() (just a passthrough)

aslo downgraded "GDB Python API is supported only for local processes" to a warning, because it seems to work on some systems.

Commit:
added changelog entry and fixed python2.7 compatibility

Commit:
also added optional gdbserver_args parameter to gdb.debug()

gfelber commented:
additionally added the optional argument `gdbserver_args`, which could be useful for some setups or troubleshooting

peace-maker commented on pwnlib/gdb.py:
```diff
@@ -628,15 +638,15 @@ def debug(args, gdbscript=None, exe=None, ssh=None, env=None, sysroot=None, api=
-            args = _gdbserver_args(args=args, which=which, env=env, python_wrapper_script=script)
+            args = _gdbserver_args(gdbserver_args=gdbserver_args, args=args, port=port, which=which, env=env, python_wrapper_script=script)
     else:
         qemu_port = random.randint(1024, 65535)
```
We should use the provided port here too if it is set. Right now qemu listens on a different port than gdb is targeting below.

gfelber replied:
7305502: gdb.debug port is now also used for qemu

peace-maker commented on pwnlib/gdb.py:
```diff
@@ -612,7 +622,7 @@ def debug(args, gdbscript=None, exe=None, ssh=None, env=None, sysroot=None, api=
 
     if api and runner is not tubes.process.process:
-        raise ValueError('GDB Python API is supported only for local processes')
+        log.warn('GDB Python API is supported only for local processes')
```
I don't like changing this without tests that proove full support. Then we could just remove the warning altogether. Which runner did work for you other than a process?

gfelber replied:
ssh.process works for me.
This makes sense IMO, because gdb is still run on the host session therefore attaching the gdb api through a unix socket still works. Changing it into a warning for ssh.process and an error for everything else

gfelber replied:
7305502: GDB Python API is now a warning for ssh.process and an error for everything else

peace-maker commented on pwnlib/gdb.py:
```diff
@@ -412,17 +419,20 @@ def _get_runner(ssh=None):
         exe(str): Path to the executable on disk
         env(dict): Environment to start the binary in
         ssh(:class:`.ssh`): Remote ssh session to use to launch the process.
+        port(int): Gdb port to use
```
Maybe mention it using a random one if the argument is unchanged.

gfelber replied:
7305502: updated docs to use mention that gdbserver ports are randomized by default

peace-maker commented on pwnlib/gdb.py:
```diff
@@ -667,17 +677,19 @@ def debug(args, gdbscript=None, exe=None, ssh=None, env=None, sysroot=None, api=
+    if not port:
+        # Find what port we need to connect to
+        if ssh or context.native or (context.os == 'android'):
+            port = _gdbserver_port(gdbserver, ssh)
```
I think we should still parse the output even though the port is specified explicitly. There were problems with shells forking when gdbserver uses them causing stalls which were detected by a timeout on the message in `_gdbserver_port` recently. And this way we're sure gdbserver is up and running before trying to attach.

gfelber replied:
7305502: now using gdbserver_port to check if the correct port was set

peace-maker commented on CHANGELOG.md:
```diff
@@ -71,6 +71,7 @@ The table below shows which release corresponds to each branch, and what date th
 
 ## 4.14.0 (`dev`)
 
+- [#2382][2382] added optional port, gdb_args and gdbserver_args parameters to gdb.debug()
```
Please always add to the bottom of the list and add the link to the PR below.

gfelber replied:
7305502: fixed CHANGELOG.md structure

peace-maker reviewed (changes requested):
Cool, thank you! Allowing more control over the way debugging is initiated seems useful for one-off problems. 

Your "plugin" looks nice, do you have plans to upstream features? I'd rather have nice docker and qemu integration than patchwork changes to make the external tool work. :D

gfelber committed:
Merge branch 'Gallopsled:dev' into dev

gfelber committed:
implemented requested fixes for #2382:
+ gdb.debug port is now also used for qemu
+ GDB Python API is now tested for tubes.process.process a warning for ssh.process and an error for everything else
+ updated docs to use mention that gdbserver ports are randomized by default
+ now using gdbserver_port to check if the correct port was set
+ fixed CHANGELOG.md structure
+ added timeouts for recvline() gdb tests
+ also run gdb api tests for ssh runner

gfelber committed:
implemented requested fixes for #2382:
+ gdb.debug port is now also used for qemu
+ GDB Python API is now tested for tubes.process.process a warning for ssh.process and an error for everything else
+ updated docs to use mention that gdbserver ports are randomized by default
+ now using gdbserver_port to check if the correct port was set
+ fixed CHANGELOG.md structure

gfelber commented:
Addressed change requests in 7305502.

regarding:
> Your "plugin" looks nice, do you have plans to upstream features? I'd rather have nice docker and qemu integration than patchwork changes to make the external tool work. :D

The initial reason this has turned into a "plugin" i maintain myself is so i can basically make changes whenever i encounter a problem or run into an issue (or want to add a new feature).

Generally speaking I don't mind adding the functionality to upstream, but this might require some time because
+ i will have to rewrite a lot of stuff to work with python2.7 
+ i will have to read into the pwntools project to understand exactly how your docs, code structure and testing setups work
I would probably also keep my repo around, so it's easier for me to add and test new stuff

Maybe we can move this discussion somewhere else? My discord tag is @gfelber (I'm also on the official discord server).

peace-maker commented on pwnlib/gdb.py:
```diff
@@ -616,7 +626,9 @@ def debug(args, gdbscript=None, exe=None, ssh=None, env=None, sysroot=None, api=
-        raise ValueError('GDB Python API is supported only for local processes')
+        if runner is not ssh.process:
+            raise ValueError('GDB Python API is supported only for local processes')
+        log.warn('GDB Python API for ssh processes is not officially tested')
```
It would be better to add the tests instead of this vague warning. I think we can copy the API tests and slap a `ssh=shell` argument to the `.debug` call.

See [here for instructions](https://github.com/Gallopsled/pwntools/blob/dev/travis/docker/README.md) on how to test locally. `make -C travis/docker ANDROID=no TARGET=gdb.rst` should do the job.

peace-maker replied:
Ah, lol. gdb tests are disabled in the docker testing setup apparently. CI will do the work then.
https://github.com/Gallopsled/pwntools/blob/51e8eb09ee206de670e688ea5801f259cd8b80e4/travis/docker/doctest3#L27-L29

peace-maker replied:
I just tried running the gdb tests in docker and they all passed, so I guess we can just remove that line from doctest3 and include them.

gfelber replied:
0ee8d36  so i added some test cases (the same as non ssh) and i was wondering how this test case is supposed to work (IMO it doesn't make sense for gdb.attach to work on remote targets and the test case doesn't work for me).

```python
        Attach to the remote process from a :class:`.remote` or :class:`.listen` tube,
        as long as it is running on the same machine.

        >>> server = process(['socat', 'tcp-listen:12345,reuseaddr,fork', 'exec:/bin/bash,nofork'])
        >>> sleep(1) # Wait for socat to start
        >>> io = remote('127.0.0.1', 12345)
        >>> sleep(1) # Wait for process to fork
        >>> pid = gdb.attach(io, gdbscript='''
        ... call puts("Hello from remote debugger!")
        ... detach
        ... quit
        ... ''')
        >>> io.recvline()
        b'Hello from remote debugger!\n'
        >>> io.sendline(b'echo Hello from bash && exit')
        >>> io.recvall()
        b'Hello from bash\n'
```

this is the error i get

```
Document: gdb
-------------
**********************************************************************
File "../../pwnlib/gdb.py", line ?, in default
Failed example:
    io.recvline()
Exception raised:
    Traceback (most recent call last):
      File "/usr/lib/python3.10/doctest.py", line 1350, in __run
        exec(compile(example.source, filename, "single",
      File "<doctest default[5]>", line 1, in <module>
        io.recvline()
      File "/home/pwntools/pwntools/pwnlib/tubes/tube.py", line 527, in recvline
        return self.recvuntil(self.newline, drop = not keepends, timeout = timeout)
      File "/home/pwntools/pwntools/pwnlib/tubes/tube.py", line 341, in recvuntil
        res = self.recv(timeout=self.timeout)
      File "/home/pwntools/pwntools/pwnlib/tubes/tube.py", line 106, in recv
        return self._recv(numb, timeout) or b''
      File "/home/pwntools/pwntools/pwnlib/tubes/tube.py", line 176, in _recv
        if not self.buffer and not self._fillbuffer(timeout):
      File "/home/pwntools/pwntools/pwnlib/tubes/tube.py", line 155, in _fillbuffer
        data = self.recv_raw(self.buffer.get_fill_size())
      File "/home/pwntools/pwntools/pwnlib/tubes/sock.py", line 39, in recv_raw
        data = self.sock.recv(numb, *a)
      File "/home/pwntools/pwntools/docs/source/conf.py", line 446, in alrm_handler
        raise EndlessLoop()
    EndlessLoop

```

i also added some timeouts to the other recvline commands for the tests

gfelber committed:
Merge branch 'dev' into dev

gfelber commented:
hey are there any outstanding issues?
Without this patch I can't start with up streaming [vagd](https://github.com/gfelber/vagd)

peace-maker reviewed (approved):
No, sorry. I wanted to check why you had to add the timeout everywhere instead of having a timeout set on the context but didn't get to it.
I can clean up later if there is something to cleanup.

Thank you!

peace-maker committed:
Merge branch 'dev' into dev

gfelber commented:
> No, sorry. I wanted to check why you had to add the timeout everywhere instead of having a timeout set on the context but didn't get to it.

Zeah i guess context would have been better for the docs, but i realized that there are different acceptable timeout margins (e.g. some use 1s others 30s), so decided to set it to 1s for most use cases and in special cases increased it
